rfd/XXXX-upgrade-system.md

```diff
@@ -0,0 +1,659 @@
+be periodically generated by teleport or by some custom external program. It may
+also be manually created by an administrator.
+
+- `version-controller`: A pluggable component responsible for generating the
+`version-directive` based on some dynamic state (e.g. a server which publishes
+package versions and hashes). A builtin `version-controller` would simply be a
+configuration resource from the user's perspective. A custom/external version
+controller would be any program with the permissions necessary to update the
+`version-directive` resource.
+
+- *version reconciliation loop*: A control loop that runs in the auth server which
+comparse the desired state as specified by the `version-directive` with the current
+state of teleport installations across the cluster. When mismatches are discovered,
+the appropriate `install-controller`s are run.
+
+- `install-controller`: A component capable of attempting to affect installation
+of a specific target on one or more teleport hosts. The auth server needs to know
+enough to at least start the process, but the core logic of a given `install-controller`
+may be an external command (e.g. the proposed `local-script` controller would cause
+each teleport instance in need of upgrade to run a user-supplied script locally and
+then restart). From the perspective of a user, an `install-controller` is a teleport
+configuration object (though that configuration object may only be a thin hook into
+the "real" `install-controller`). Whether or not the teleport instance being upgraded
+needs to understand the `install-controller` will depend on the specific controller.
+
+There is room in the above model for additional granularity (e.g. schedulers), but
+the listed components are a decent starting point.
+
+### Implementation Plan Overview
+
+Implementation will consist of four main phases:
+
+1. Setup: Refactor our various service heartbeats into a unified bidirectional
+GRPC stream. This will allow us to track the existence of teleport instances in a
+consistent manner regardless of the specific services running on a given node.
+It will also give us a mechanism for directly signaling individual nodes in order
+to perform rolling upgrades. Improve teleport's self-knowledge, so that instances
+can accurately report the specific build parameters that they correspond to
+(arch/fips/etc) in addition to version. Make version inventory info available
+via `tctl`.
+
+2. MVP: Implement the `version-directive` resource, version reconciliation loop,
+and `local-script` installer. This will serve both as a pathfinder/poc,
+and also allow us to deliver a working means of easily performing cluster-wide
+upgrades via existing upgrade mechanisms (e.g. `apt`/`yum`) in a much shorter time
+than if we waited on the full "batteries included" feature being finished. This
+iteration will only support non-auth instances, and remote rollbacks.
+
+3. Batteries Included: Implement a `version-controller` and `install-controller`
```

Comment:
Can you say a couple of words about future K8s controller as well? Afaik this should map perfectly on K8s controllers, but wanted to make sure.